src/librustc_typeck/check/method/probe.rs

```diff
@@ -962,28 +1031,45 @@ impl<'a, 'gcx, 'tcx> ProbeContext<'a, 'gcx, 'tcx> {
         let region = tcx.mk_region(ty::ReErased);
 
         // Search through mutabilities in order to find one where pick works:
-        [hir::MutImmutable, hir::MutMutable]
-            .iter()
-            .filter_map(|&m| {
+        let mut elements = [hir::MutImmutable, hir::MutMutable];
+        let mut it = elements
+            .iter_mut()
+            .filter_map(|&mut m| {
                 let autoref_ty = tcx.mk_ref(region,
                                             ty::TypeAndMut {
                                                 ty: step.self_ty,
                                                 mutbl: m,
                                             });
                 self.pick_method(autoref_ty).map(|r| {
-                    r.map(|mut pick| {
-                        pick.autoderefs = step.autoderefs;
-                        pick.autoref = Some(m);
-                        pick.unsize = if step.unsize {
-                            Some(step.self_ty)
-                        } else {
-                            None
-                        };
-                        pick
+                    r.map(|mut picks| {
+                        for pick in picks.iter_mut() {
+                            pick.autoderefs = step.autoderefs;
+                            pick.autoref = Some(m);
+                            pick.unsize = if step.unsize {
+                                Some(step.self_ty)
+                            } else {
+                                None
+                            };
+                        }
+                        picks
                     })
                 })
-            })
-            .nth(0)
+            });
+        match self.looking_for {
+            LookingFor::MethodName(_) => it.nth(0),
```

Comment:
Let's add an assertion here that there is exactly one; not sure the most elegant way to do that.